core/directory.go

```diff
@@ -0,0 +1,291 @@
+package core
+
+import (
+	"context"
+	"path"
+
+	"github.com/moby/buildkit/client/llb"
+	bkgw "github.com/moby/buildkit/frontend/gateway/client"
+	"github.com/moby/buildkit/solver/pb"
+	"go.dagger.io/dagger/core/schema"
+	"go.dagger.io/dagger/router"
+)
+
+// Directory is a content-addressed directory.
+type Directory struct {
+	ID DirectoryID `json:"id"`
+}
+
+// DirectoryID is an opaque value representing a content-addressed directory.
+type DirectoryID string
+
+// directoryIDPayload is the inner content of a DirectoryID.
+type directoryIDPayload struct {
+	LLB *pb.Definition `json:"llb"`
+	Dir string         `json:"dir"`
+}
+
+func (id DirectoryID) decode() (*directoryIDPayload, error) {
+	var payload directoryIDPayload
+	if err := decodeID(&payload, id); err != nil {
+		return nil, err
+	}
+
+	return &payload, nil
+}
+
+func NewDirectory(ctx context.Context, st llb.State, cwd string) (*Directory, error) {
+	def, err := st.Marshal(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	id, err := encodeID(directoryIDPayload{
+		LLB: def.ToPB(),
+		Dir: cwd,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	return &Directory{
+		ID: DirectoryID(id),
+	}, nil
+}
+
+func (dir *Directory) Contents(ctx context.Context, gw bkgw.Client, src string) ([]string, error) {
+	st, cwd, err := dir.Decode()
+	if err != nil {
+		return nil, err
+	}
+
+	def, err := st.Marshal(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	res, err := gw.Solve(ctx, bkgw.SolveRequest{
+		Definition: def.ToPB(),
+	})
+	if err != nil {
+		return nil, err
+	}
+	ref, err := res.SingleRef()
+	if err != nil {
+		return nil, err
+	}
+
+	if ref == nil {
+		// empty directory, i.e. llb.Scratch()
+		return []string{}, nil
+	}
+
+	entries, err := ref.ReadDir(ctx, bkgw.ReadDirRequest{
+		Path: path.Join(cwd, src),
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	paths := []string{}
+	for _, entry := range entries {
+		paths = append(paths, entry.GetPath())
+	}
+
+	return paths, nil
+}
+
+func (dir *Directory) WithNewFile(ctx context.Context, gw bkgw.Client, dest string, content []byte) (*Directory, error) {
+	st, cwd, err := dir.Decode()
+	if err != nil {
+		return nil, err
+	}
+
+	// be sure to create the file under the working directory
+	dest = path.Join(cwd, dest)
```

Comment:
My expectation would have been that if I provide a relative `dest` then it should be relative to `cwd`, but if `dest` is absolute then it should just be what I tell it. Not sure if that's aligned with everyone else's expectations, so not a blocker if others disagree.

Reply:
I'm trying to form an opinion on this, but don't understand what `cwd` is. Seems like an implementation detail?

Reply:
This line is to handle this possibly-edge case:

```graphql
directory {
  withNewFile(path: "foo/bar", contents: "bar") {
    directory(path: "foo") {
      withNewFile(path: "baz", contents: "baz")
    }
  }
}
```

The intention above is to create this file tree, since I did the `withNewFile` relative to the inner `directory {}`:

```
foo/bar
foo/baz
```

But without this line it would end up creating it in the root of the outer directory:

```
foo/bar
baz
```

And I think what @sipsma is suggesting is that if I explicitly pass an absolute path, like this:


```graphql
directory {
  withNewFile(path: "foo/bar", contents: "bar") {
    directory(path: "foo") {
      withNewFile(path: "/baz", contents: "baz")
    }
  }
}
```

...then it should end up with the above result, because I explicitly told it to create the file in the root of the directory.

```
foo/bar
baz
```

I think that makes sense from the standpoint of respecting the path syntax, but I wonder if it'll be hard to tell where exactly the "root" is in practice. :thinking: For example, if I take the `id` of the inner directory and hop into it from a context-free `directory(id: $id)` it feels a little weird to allow the user to create files outside of it.

```graphql
directory(id: $subdir) {
  withNewFile(path: "/absolute", content: "abs") {
    id
  }
}
```

I suppose you'll still get an `ID` back with the outer directory modified, but I don't know how you would gain access to the file you created, unless we add an accessor to the parent dir.

Reply:
Ohhhh sorry, I totally misunderstood what `cwd` was actually. I just instantly assumed it was the implementation of "workdir" since that's usually what I think of when I see `cwd`, but that doesn't even make sense in `Directory` yet.

I agree that this should always be relative to the subdir (aka `cwd` var here) no matter whether it's relative or absolute. We should probably treat going to a subdir of a `Directory` spiritually the same as a `chroot`, shouldn't be able to access any files above that subdir once it's been selected.

Reply:
> We should probably treat going to a subdir of a Directory spiritually the same as a chroot, shouldn't be able to access any files above that subdir once it's been selected.

Agreed!